impl/src/as/mod.rs

```diff
@@ -178,92 +184,120 @@ impl<'a> ToTokens for Expansion<'a> {
+                true,
+                Either::Right(tys.iter().map(ToTokens::into_token_stream)),
+            ),
+            None => (
+                false,
+                false,
+                Either::Left(std::iter::once(quote! { #field_ty })),
+            ),
+        };
+
+        return_tys
+            .map(|return_ty| {
+                let trait_ty = quote! { ::core::convert::#trait_ident <#return_ty> };
+
+                let generics = if forward_impl {
+                    let mut generics = self.generics.clone();
+                    if blanket {
+                        generics.params.push(parse_quote! { #return_ty });
+                        generics
+                            .make_where_clause()
+                            .predicates
+                            .push(parse_quote! { #return_ty: ?::core::marker::Sized });
+                    }
+                    generics
+                        .make_where_clause()
+                        .predicates
+                        .push(parse_quote! { #field_ty: #trait_ty });
+
+                    Cow::Owned(generics)
+                } else {
+                    Cow::Borrowed(self.generics)
+                };
+
+                let (impl_gens, _, where_clause) = generics.split_for_impl();
+                let (_, ty_gens, _) = self.generics.split_for_impl();
+
+                let mut body = quote! { & #mut_ self.#field_ident };
+                if forward_impl {
+                    body = quote! { <#field_ty as #trait_ty>::#method_ident(#body) };
+                }
+
+                quote! {
+                    #[automatically_derived]
+                    impl #impl_gens #trait_ty for #ty_ident #ty_gens #where_clause {
+                        #[inline]
+                        fn #method_ident(& #mut_ self) -> & #mut_ #return_ty {
+                            #body
+                        }
+                    }
+                }
+            })
+            .collect::<TokenStream>()
+            .to_tokens(tokens);
     }
 }
 
 /// Representation of an [`AsRef`]/[`AsMut`] derive macro struct container attribute.
 ///
 /// ```rust,ignore
 /// #[as_ref(forward)]
+/// #[as_ref(<types>)]
 /// ```
-type StructAttribute = forward::Attribute;
+type StructAttribute = AsArgs;
 
 /// Representation of an [`AsRef`]/[`AsMut`] derive macro field attribute.
 ///
 /// ```rust,ignore
 /// #[as_ref]
 /// #[as_ref(forward)]
+/// #[as_ref(<types>)]
 /// #[as_ref(skip)] #[as_ref(ignore)]
 /// ```
 enum FieldAttribute {
     Empty,
-    Forward(forward::Attribute),
+    Args(AsArgs),
     Skip(skip::Attribute),
 }
 
+/// Arguments specifying which conversions should be generated
+///
+/// ```rust,ignore
+/// #[as_ref(forward)]
+/// #[as_ref(<types>)]x
+/// ```
+enum AsArgs {
+    /// Blanket impl, fully forwarding to the field type
+    Forward(forward::Attribute),
+    /// Forward implementation, but only impl for specified types
+    Types(Punctuated<Type, Token![,]>),
```

Comment:
I don't think it makes sense to extract the type list attribute into `utils` like forward/skip, even though
the same arguments are used in `From` and `Into`, because parsing it is quite concise anyway